swingset/src/main/java/com/nqadmin/swingset/formatting/SSFormattedTextField.java

```diff
@@ -86,59 +86,76 @@
 		implements FocusListener, SSComponentInterface {
 
 	/**
-	 * This class should implements validation AND calls setValue() which triggers RowSet updates.
+	 * We want an InputVerifier in order to lock the focus down while the JFormattedTextField is in
+	 * an invalid edit state.
+	 * 
+	 * Also, we add a call to validateField(), the default implementation of which just returns true.
+	 * This allows for the developer to add additional checks (normally range validation) beyond what
+	 * is provided by the Formatter/FormatterFactory, which generally only handles display of values
+	 * and/or character masks.
+	 * 
+	 * This class should perform validation AND call setValue() which will trigger a RowSet update.
 	 * <p>
-	 * See https://docs.oracle.com/javase/8/docs/api/javax/swing/JFormattedTextField.html
-	 * <p>
-	 * More on input verifiers here: https://www.drdobbs.com/jvm/java-better-interfaces-via-jformattedtex/224700979?pgno=1
-	 * Some discussion on dates and validation here: https://docs.oracle.com/javase/tutorial/uiswing/misc/focus.html#inputVerification
+	 * See https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/JFormattedTextField.html
 	 */
-	// TODO: This still needs some work.
 	public class FormattedTextFieldVerifier extends InputVerifier {
 
 		@Override
 		public boolean verify(final JComponent input) {
 
 			boolean result = true;
 			Object value = null;
+			
+			// Set indicator to suppress duplicate property change event if
+			// a call is made to setValue()
+			verifyingText = true;
 
 			if (input instanceof SSFormattedTextField) {
 				logger.debug("{}: Instance of SSFormattedTextField.", () -> getColumnForLog());
+				
 				final SSFormattedTextField ssftf = (SSFormattedTextField) input;
-				AbstractFormatter formatter = ssftf.getFormatter();
 				String formattedText = ssftf.getText();
+				
+				AbstractFormatter formatter = ssftf.getFormatter();
 				logger.debug("Formatter is: " + formatter + ".");
+				if (formatter==null) {
+					logger.warn("Null formatter encountered for formatted text field.");
+				}
```

Comment:
I think it would be a programmer error.

Agree that it's mild, but we were not giving any sort of error/warning before. I had added a TODO to look at this because taking no action at all seemed insufficient. Per the 17 JavaDoc, `If an AbstractFormatterFactory has not been explicitly set, one will be set based on the Class of the value type after setValue has been invoked (assuming value is non-null).` I guess developer could instantiate a SSFormattedText field with null or explicitly set the formatter to null. We could present a popup to the end user, but if we have formatter set to null then either we screwed up or the app developer screwed up.

`logger.error(` or we could throw an exception?

Reply:
Exception? IIRC, SwingSet avoids exceptions even for programmer errors. Although, in some cases, SQL exception can be thrown since it will get caught, dialog report, and get you back to a known state. If it's not caught, probably don't want to bomb the app. If you've got automated log scanners then error should be sufficient?

Reply:
Went with `logger.error("Null formatter encountered for formatted text field.");`